introduced MPI version of nma_aligmnemt_vol program

MohamadHarastani opened the pull request:
This version was tested at IMPMC/Paris.

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+/***************************************************************************
+ *
+ * Authors:  Slavica Jonic slavica.jonic@impmc.jussieu.fr
```
Who is the author? Slavica Jonic or you?

MohamadHarastani replied:
This was written by both of us, I forgot to add my name.

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+class MpiProgNMAVol: public ProgNmaAlignmentVol
+{
+private:
+    MpiNode *node;
```
follow[ rule of 5 or 0](https://en.cppreference.com/w/cpp/language/rule_of_three)

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+/** Class to perfom the NMA Alignment with  MPI parallelization */
+class MpiProgNMAVol: public ProgNmaAlignmentVol
+{
+private:
```
Extract the header file

MohamadHarastani replied:
By extract, do you mean remove?

DStrelak replied:
no, I mean extract (one .h file with declarations, one .cpp file with definitions)

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+    /** Empty Constructor */
+    MpiProgNMAVol()
+    {
+    	node=NULL;
```
`nullptr`

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+        if (node->isMaster())
+        {
+            ProgNmaAlignmentVol::createWorkFiles();
+            mdIn.write(fnOutDir + "/nmaTodo.xmd");
```
`/nmaTodo.xmd` should be a constant.
Is it 'To-Do' (eng) or 'todo' (esp)? Consider using a better name

MohamadHarastani replied:
It is To-Do in English. This file is temporary (gets deleted by the end of the protocol), and allows to 'continue' from one point in case the program stops (sudden restarts for instance). The people that use it know where to find it and what is it's name. So I prefer to keep it as it is.

DStrelak replied:
Ok, so extract the name as a constant

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+        node->barrierWait();//Sync all before start working
+        mdIn.read(fnOutDir + "/nmaTodo.xmd");
+        mdIn.findObjects(imgsId);//get objects ids
+        rangen = node->rank;
```
What is rangen? From the description it's a random generator seed, but it's never used that way. Please fix it.

MohamadHarastani replied:
The name originates to nma_alignment.cpp and nma_alignment_vol.cpp and the authors had a justification for the name. If at some point this name changes over there, it would follow here. For the moment, let's keep it as it is.

DStrelak replied:
in `nma_alignment.cpp` it is never used as a random generator seed. Please rename it approprietly

DStrelak commented on src/xmipp/applications/programs/mpi_nma_alignment_vol/mpi_nma_alignment_vol_main.cpp:
```diff
@@ -0,0 +1,140 @@
+
+        if (moreTasks)
+        {
+            time_bar_done = first + 1;
```
I sense a race condition here. Can you check it?

MohamadHarastani replied:
You may be true, I don't see the progress of the program when mpi is used. This was never reported as an issue because the time_bar doesn't effect the functionality of the program.
Do you have suggestions on how to solve it if exists? Thanks.

DStrelak replied:
Have a look [here](https://www.modernescpp.com/index.php/prefer-locks-to-mutexes) and e.g. [here](https://en.cppreference.com/w/cpp/thread/unique_lock)

DStrelak replied:
But don't worry about this one, we should deal with it properly at the level of all Xmipp programs

MohamadHarastani commented:
What exists is an mpi version for nma_alignment, this was done for nma_alignment_vol.
The program was written exactly in the same way the mpi version of nma_alignmemt was in Scipion 1, and was tested in Scipion 2 recently and workd properly. However, we will review these changes.

MohamadHarastani commented:
Hi, I have tracked back all the points to this file you pointed to this file [https://github.com/I2PC/xmipp/blob/devel/src/xmipp/libraries/parallel/mpi_nma_alignment.cpp](url)
All your suggestions applies to this file, and I was not planning to touch it. What do you suggest?

DStrelak commented:
Well, that happens when you copy and paste :)
The easiest way might be to create a common implementation for the MPI version (sth like `MPINmaAlignment`), and in the `MpiProgNMA` \ `MpiProgNMAVol` you inherit from both (`MPINmaAlignment` and `ProgNMAVol`) and call methods via `this`.

MohamadHarastani commented:
Hello, I need some time before you make another review. Thanks for waiting.

MohamadHarastani commented:
I did the majority of the changes and restructured files (main, and one .h another .cpp files). I couldn't follow rule zero (the program needs a destructor) or five (it is complicated to make proper definition of the other functions because they are not used and I don't know what to do), I would really appreciate some help from your side.
I have also introduced a new option to xmipp_volume_align that stores the parameters (shifts&angles) that the program finds during the alignment (I also reverted what I opened the issue report about [here](https://github.com/I2PC/xmipp/issues/213)). These parameters can be later used to align the volumes (by xmipp_transform_geometry). And introduced some lines to (nma_alignment_vol.cpp) that uses this last option (of storing those rigid body alignment parameters) besides the elastic alignment parameters.
Thanks.

DStrelak commented on src/xmipp/libraries/parallel/mpi_nma_alignment_vol.cpp:
```diff
@@ -0,0 +1,125 @@
+ *  e-mail address 'xmipp@cnb.uam.es'
+ ***************************************************************************/
+#include <parallel/xmipp_mpi.h>
+#include <reconstruction/nma_alignment_vol.h>
```
first two includes can be in the header

DStrelak commented on src/xmipp/libraries/parallel/mpi_nma_alignment_vol.h:
```diff
@@ -0,0 +1,69 @@
+class MpiProgNMAVol: public ProgNmaAlignmentVol
+{
+private:
+    MpiNode *node;
```
How about [smart pointers](https://en.cppreference.com/w/cpp/memory/shared_ptr) ?

MohamadHarastani replied:
I believe that your idea is to use smart pointers to get rid of the destructor and use rule of zero. I followed the link and tried to keep it simple, but I faced difficulties.
Lets say that all of **node**, **distributor** and **fileMutex** are smart pointers. Then such a command:
"fileMutex = new MpiFileMutex(node)" replaced by "fileMutex = std::make_shared<MpiFileMutex>(node)" doesn't work in principle, because the class MpiFileMutex expects a raw pointer node not a shared pointer. Or at least this is what I could get from the error while compiling.

DStrelak replied:
Of course, you would have to change the type of the variables.
To access the raw pointer, you can use get() method

MohamadHarastani replied:
Thanks for the hint. All works well now!

DStrelak commented on src/xmipp/libraries/parallel/mpi_nma_alignment_vol.h:
```diff
@@ -0,0 +1,69 @@
+    void read(int argc, char **argv);
+    
+    // main body
+    void createWorkFiles();
```
Do all these have to be `public`?

MohamadHarastani replied:
Except of read() they can be private. I will move them.

DStrelak commented on src/xmipp/libraries/reconstruction/nma_alignment_vol.cpp:
```diff
@@ -207,11 +217,14 @@ FileName ProgNmaAlignmentVol::createDeformedPDB() const {
 		trial_best = trial;
+		return true;
 	}
+	else
```
no need for else, simply `return false`

DStrelak commented on src/xmipp/libraries/reconstruction/nma_alignment_vol.cpp:
```diff
@@ -224,10 +237,15 @@ double ObjFunc_nma_alignment_vol::eval(Vector X, int *nerror) {
 	FileName fnRandom = global_nma_vol_prog->createDeformedPDB();
 	const char * randStr = fnRandom.c_str();
+	FileName fnShiftsAngles = formatString("%s_angles_shifts.txt", randStr);
+	const char * shifts_angles = fnShiftsAngles.c_str();
```
This is already getting confusing. 
Isn't it simply `fnRandom + "_angles_shifts.txt"`?
